linker-interface/shared/src/main/scala/org/scalajs/linker/interface/Fingerprint.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.interface
+
+import scala.collection.mutable
+
+private[interface] trait Fingerprint[T] {
+  def fingerprint(obj: T): String
+}
+
+private[interface] object Fingerprint {
+  def fingerprint[T: Fingerprint](obj: T): String =
+    implicitly[Fingerprint[T]].fingerprint(obj)
+
+  final class FingerprintBuilder(name: String) {
+    private val fields = new mutable.ListBuffer[String]
+
+    def addField[T: Fingerprint](name: String, value: T): FingerprintBuilder = {
+      fields.append(s"$name=${fingerprint(value)}")
+      this
+    }
+
+    def build(): String = fields.mkString(s"$name(", ",", ")")
+  }
+
+  implicit object BooleanFingerprint extends Fingerprint[Boolean] {
+    override def fingerprint(bool: Boolean): String = bool.toString
+  }
+
+  implicit object IntFingerprint extends Fingerprint[Int] {
+    override def fingerprint(int: Int): String = int.toString
+  }
+
+  implicit object StringFingerprint extends Fingerprint[String] {
+    override def fingerprint(str: String): String = str
+  }
+
+  implicit def optionFingerprint[T: Fingerprint]: Fingerprint[Option[T]] = {
+    new Fingerprint[Option[T]] {
+      override def fingerprint(opt: Option[T]): String = {
+        opt
+          .map(implicitly[Fingerprint[T]].fingerprint)
+          .fold("None")(f => s"Some($f)")
+      }
+    }
+  }
+
+  implicit def listFingerprint[T: Fingerprint]: Fingerprint[List[T]] = {
+    new Fingerprint[List[T]] {
+      override def fingerprint(list: List[T]): String = {
+        list
+          .map(implicitly[Fingerprint[T]].fingerprint)
```

Comment:
Just saw this: Simply `.map(fingerprint(_))`?

Reply:
I'm not sure I follow. How would this type-check? The `fingerprint` I need to apply to list items is the one for `T`, not `List[T]`. 

Would you prefer the alternative notation that takes an `implicit itemFingerprint: Fingerprint[T]` instead of a context bound `[T: Fingerprint]` and `implicitly[Fingerprint[T]]`?

Reply:
Ah, ah, I see the problem: I was referring to the top-level fingerprint that merely calls `fingerprint` on its own implicit parameter. But because of the same name, that of course doesn't work, because `fingerprint` refers to the `def` itself. It's just a nuisance anyways. So maybe fine to leave it as such.